backend/src/openai.ts

```diff
@@ -333,108 +374,64 @@ function pushCompletionToHistory(
 	const maxChatHistoryLength = 1000;
+	const updatedChatHistory = [...chatHistory];
 
 	if (chatMessageType !== CHAT_MESSAGE_TYPE.BOT_BLOCKED) {
 		// remove the oldest message, not including system role message
 		if (chatHistory.length >= maxChatHistoryLength) {
 			if (chatHistory[0].completion?.role !== 'system') {
-				chatHistory.shift();
+				updatedChatHistory.shift();
 			} else {
-				chatHistory.splice(1, 1);
+				updatedChatHistory.splice(1, 1);
 			}
 		}
-		chatHistory.push({
+		updatedChatHistory.push({
 			completion,
 			chatMessageType,
 		});
 	} else {
 		// do not add the bots reply which was subsequently blocked
 		console.log('Skipping adding blocked message to chat history', completion);
 	}
-	return chatHistory;
-}
-
-function getBlankChatResponse(): ChatResponse {
-	return {
-		completion: null,
-		defenceReport: {
-			blockedReason: '',
-			isBlocked: false,
-			alertedDefences: [],
-			triggeredDefences: [],
-		},
-		wonLevel: false,
-		openAIErrorMessage: null,
-	};
-}
-
-function applyOutputFilterDefence(
-	message: string,
-	defences: Defence[],
-	chatResponse: ChatResponse
-) {
-	const detectedPhrases = detectFilterList(
-		message,
-		getFilterList(defences, DEFENCE_ID.FILTER_BOT_OUTPUT)
-	);
-
-	if (detectedPhrases.length > 0) {
-		console.debug(
-			'FILTER_BOT_OUTPUT defence triggered. Detected phrases from blocklist:',
-			detectedPhrases
-		);
-		if (isDefenceActive(DEFENCE_ID.FILTER_BOT_OUTPUT, defences)) {
-			chatResponse.defenceReport.triggeredDefences.push(
-				DEFENCE_ID.FILTER_BOT_OUTPUT
-			);
-			chatResponse.defenceReport.isBlocked = true;
-			chatResponse.defenceReport.blockedReason =
-				'Message Blocked: My response was blocked as it contained a restricted word/phrase.';
-		} else {
-			chatResponse.defenceReport.alertedDefences.push(
-				DEFENCE_ID.FILTER_BOT_OUTPUT
-			);
-		}
-	}
+	return updatedChatHistory;
 }
 
 async function performToolCalls(
-	chatResponse: ChatResponse,
 	toolCalls: ChatCompletionMessageToolCall[],
 	chatHistory: ChatHistoryMessage[],
 	defences: Defence[],
 	sentEmails: EmailInfo[],
 	currentLevel: LEVEL_NAMES
-) {
+): Promise<ToolCallResponse> {
 	for (const toolCall of toolCalls) {
 		// only tool type supported by openai is function
 
 		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
 		if (toolCall.type === 'function') {
-			// call the function and get a new reply and defence info from
 			const functionCallReply = await chatGptCallFunction(
-				chatResponse.defenceReport,
 				defences,
 				toolCall.id,
 				toolCall.function,
 				sentEmails,
 				currentLevel
 			);
-			chatResponse.wonLevel = functionCallReply.wonLevel;
-
-			// add the function call to the chat history
-			pushCompletionToHistory(
-				chatHistory,
-				functionCallReply.completion,
-				CHAT_MESSAGE_TYPE.FUNCTION_CALL
-			);
-			// update the defence info
-			chatResponse.defenceReport = functionCallReply.defenceReport;
+			return {
```

Comment:
For now the only supported tool call is function calls, so that would be the case is the case until openai extends this. I can add a comment to make this clear

Reply:
I've written an investigation ticket #764